### ISSUE TITLE
Use global network policies for basic isolation of future namespaces

### ISSUE THREAD

**Issue by perezjasonr** [projectcalico/calico#6107]
Is there a way to make a global network policy so that it can have basic rules/isolation for newly created namespaces?

essentially, id like to have something like this in my k8s cluster:

```
types:
    - Ingress
    - Egress
  ingress:
  - action: Allow
    protocol: TCP
    source:
      selector: all()
  egress:
  - action: Allow
    protocol: TCP
    destination:
      selector: all()
```

this would mean workloads within the namespace can talk to each other. However, if the spec looks like this:

```
apiVersion: projectcalico.org/v3
kind: GlobalNetworkPolicy
metadata:
  name: basic-defaults
spec:
  # selector: projectcalico.org/namespace != "kube-system"
  namespaceSelector: somelabel != "deliberatelyexcludeMyExistingNamespaces" && kubernetes.io/metadata.name != "kube-system"
```

at first glance this might work it selects newly created namespaces in the future, but wouldn't this mean my rules allow all of these to talk to each other as well (inter namespace since it selects multiple namespaces)? or does each selector: all() in the rule apply to their respective namespaces? I read the documentation on the selector section and I couldn't quite conclude whether or not this is the case.

so for example if I have the rules above, and i create two new namespaces, test, and testns, does they each get isolated individually or does default, test, and testns all allowed to talk to each other?

Is this possible to do what I'm trying to do calico policies?

Since this is not a bug I removed the boiler plate stuff. Thank you.

**Comment by coutinhop** [comment]
I think you'd need to create a NetworkPolicy for each namespace to allow pods inside it to talk to each other and not to pods from other namespaces. I tried the GlobalNetworkPolicy you mentioned (both with and without the `allow`  action):
```apiVersion: crd.projectcalico.org/v1
kind: GlobalNetworkPolicy
metadata:
  annotations:
    projectcalico.org/metadata: '{"uid":"7587de4c-38eb-4ff1-9ce3-45201dc20643","creationTimestamp":"2022-05-17T16:10:48Z"}'
  creationTimestamp: "2022-05-17T16:10:48Z"
  generation: 4
  name: default.basic-defaults
  resourceVersion: "26968"
  uid: 50f94083-01f6-44b0-adc2-6179ff4cb87a
spec:
  ingress:
  - action: Allow
    protocol: TCP
  namespaceSelector: somelabel != "deliberatelyexcludeMyExistingNamespaces" && kubernetes.io/metadata.name
    != "kube-system"
  types:
  - Ingress
```

And it either blocks pods even inside the same namespace from talking to each other or allows cross-namespace traffic.

Tagging @caseydavenport to double check if this might be possible.

**Comment by caseydavenport** [comment]
> at first glance this might work it selects newly created namespaces in the future, but wouldn't this mean my rules allow all of these to talk to each other as well (inter namespace since it selects multiple namespaces)?

Yep, that would allow all of the pods across the selected namespaces to talk to and receive traffic from any source / destination, full stop.

I think what you want is a capability like this:

```yaml
spec:
  namespaceSelector: all()
  ingress:
    - action: Allow
      source:
        namespaceSelector: shares(kubernetes.io/metadata.name)
```

^ The behavior of the "shares" selector would mean to "Select all items that have the same value for the given label". So, this would allow you to create a single GNP that by default allows within each namespace by selecting all pods, and then only allowing from other pods that share its namespace.

Unfortunately, this isn't a capability we have in Calico today but it is something on our roadmap.

**Comment by perezjasonr** [comment]
Ok I really appreciate the confirmation and explanation. Yes I think this functionality would be amazing. imagine being able to make a networkpolicy that can account for namespaces even ones created in the future. For now, I will have to accept inter-namespace effect and use the one I made. I love that you already thought of it and its on the roadmap. Good stuff.

**Comment by caseydavenport** [comment]
> but now thinking about it i might need to replicate the namespace selector in the source/destination rules themselves as well.

Yep, that's right - there's no implicit scoping of the rule selectors based on the top-level selector for GlobalNetworkPolicy. There is an implicit scoping for normal NetworkPolicy, though.

**Comment by YitzyD** [comment]
+1 on the `shares` function. It would be an incredibly useful tool for building out generic GlobalNetworkPolicies

**Comment by atomashpolskiy** [comment]
+1 on `shares`, I have a use case where it could have saved a lot of trouble by giving an ability to define one generic network policy upfront which would then be applied ad-hoc to groups of resources

**Comment by jemag** [comment]
I was wondering if the `shares` feature is still on the roadmap?

**Comment by caseydavenport** [comment]
Afraid we haven't gotten to this one yet. I would still like to see it implemented, though.

**Comment by imbstack** [comment]
Is this a reasonable contribution to try to make for an outside contributor? We've started poking around at the code and it seems doable although not a small change. Worth us digging into it a bit more? Happy to chat in the Calico Users Slack, etc if there's more detail to go over.

**Comment by fasaxc** [comment]
Thanks for the offer to work on this. We almost started working on it as part of the upstream AdminNetowrkPolicy API, but then it was dropped from the upstream ANP spec until some kinks were ironed out.  We've been working in sig-network-policy-api to try to help move that forward.

My feeling on what the API should look like has been influenced by the discussion in sig-network-policy-api:

I don't think it should just be added to the selector syntax because that means that every call to `selector.Evaluate` now needs a "self" argument and that is tough to implement across the board (we use selectors in a lot of places apart from policy).  Or, we have to fork the selector syntax even more.

I'm inclined to make it a namespace labels-only feature in v1.  All the use cases that I've see for this are namespace-based (please shout if you've got a non-namespace use case!) and it makes the implementation a lot easier due to the lack of label inheritance.  When you put a selector in a GlobalNetworkPolicy rule, it can match on pod labels or labels inherited from a namespace with a prefix; we use that internally to implement namespaceSelectors and it makes the matching logic much more complicated.

So, my proposal would be to make this dedicated fields like this:

```
- action: Allow
  source:
    sharedNamespaceLabels: ["tenant", "group"]
- action: Deny
  source:
    notSharedNamespaceLabels: ["tenant", "group"]
```

Later if we want to do pod labels too, that would be a separate set of fields and we _still_ wouldn't need to worry about label inheritance, because we'd be explicitly talking about workload labels.

As for implementation, the approach that occurs to me is to augment Felix's ActiveRulesCalculator with namespace awareness so that it knows which policies are active for pods in which namespaces, then, at its output we add a new component that "expands" any policies with same/not same label fields into one normal policy per active namespace with the same/not same matches moved to the "real" selectors.  

If you do want to take this on, please reach out on slack in `#contributors`.  Doing work in the calculation graph can be challenging because any "match" that you calculate needs to be dynamic and it needs to respond to changes in any/all of the inputs... If a namespace is added/deleted/has labels changed then the output needs to be recalculated; if a policy is added/deleted/changed then the output needs to be recalculated and you need to emit deletion events for any expanded policies that should no longer exist.  Since that is such a rich source of bugs, the calc graph has its own test harness that must be used to test new function; it sends inputs in different orders and with different groupings to "fuzz" for common types of issues.
